services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/PushStatusCollector.java

```diff
@@ -0,0 +1,231 @@
+    for (CompletableFuture<TopicPushStatus> future: resultList) {
+      TopicPushStatus pushStatus;
+      try {
+        pushStatus = future.get();
+      } catch (Exception e) {
+        LOGGER.error("Caught exception when getting future result of push status.", e);
+        continue;
+      }
+      ExecutionStatusWithDetails daVinciStatus = pushStatus.getDaVinciStatus();
+      ExecutionStatusWithDetails serverStatus = pushStatus.getServerStatus();
+      if (serverStatus == null) {
+        continue;
+      }
+      LOGGER.info(
+          "Topic server push status: {}, Da Vinci push status: {}",
+          serverStatus.getStatus(),
+          daVinciStatus.getStatus());
+      if (isOverallPushCompleted(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        pushCompletedHandler.accept(pushStatus.getTopicName());
+      } else if (isOverallPushError(serverStatus, daVinciStatus)) {
+        pushStatus.setMonitoring(false);
+        StringBuilder pushErrorDetailStringBuilder = new StringBuilder();
+        if (serverStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          pushErrorDetailStringBuilder.append("Server push error: ").append(serverStatus.getDetails()).append("\n");
+        }
+        if (daVinciStatus.getStatus().equals(ExecutionStatus.ERROR)) {
+          pushErrorDetailStringBuilder.append("Da Vinci push error: ").append(daVinciStatus.getDetails()).append("\n");
+        }
+        pushErrorHandler.accept(pushStatus.getTopicName(), pushErrorDetailStringBuilder.toString());
+      }
+    }
+  }
+
+  public void handleServerPushStatusUpdate(String topicName, ExecutionStatus executionStatus, String detailsString) {
+
+    TopicPushStatus topicPushStatus = topicToPushStatusMap.get(topicName);
+    if ((!daVinciPushStatusScanEnabled) || topicPushStatus == null) {
+      if (executionStatus.equals(ExecutionStatus.COMPLETED)) {
+        pushCompletedHandler.accept(topicName);
+      } else if (executionStatus.equals(ExecutionStatus.ERROR)) {
+        pushErrorHandler.accept(topicName, detailsString);
+      }
+    } else {
+      // Update the server topic status in the data structure and wait for async DVC status scan thread to pick up.
+      topicToPushStatusMap.computeIfPresent(topicName, (topic, pushStatus) -> {
+        pushStatus.setServerStatus(new ExecutionStatusWithDetails(executionStatus, detailsString));
+        return pushStatus;
+      });
+    }
+  }
+
+  private boolean isOverallPushCompleted(
+      ExecutionStatusWithDetails serverStatus,
+      ExecutionStatusWithDetails daVinciStatus) {
+    return serverStatus.getStatus().equals(ExecutionStatus.COMPLETED)
+        && daVinciStatus.getStatus().equals(ExecutionStatus.COMPLETED);
+  }
+
+  private boolean isOverallPushError(
+      ExecutionStatusWithDetails serverStatus,
+      ExecutionStatusWithDetails daVinciStatus) {
+    return serverStatus.getStatus().equals(ExecutionStatus.ERROR)
```

Comment:
Shall we do a `null` check against the params since the `TopicPushStatus` doesn't init these variables at construction time?
```
    private ExecutionStatusWithDetails serverStatus;
    private ExecutionStatusWithDetails daVinciStatus;
```

Reply:
Maybe we should initialize these two variables as `NEW`?

Reply:
The check is actually done right before where it is called. I am wondering if this is actually confusing, if so I can actually fold it back to the caller method to make it more clear..